langchain/vectorstores/azuresearch.py

```diff
@@ -0,0 +1,484 @@
+    SearchIndex,
+    SemanticConfiguration,
+    SemanticField,
+    SemanticSettings,
+    SimpleField,
+    VectorSearch,
+    VectorSearchAlgorithmConfiguration,
+)
+from azure.search.documents.models import Vector
+from pydantic import BaseModel, root_validator
+
+from langchain.docstore.document import Document
+from langchain.embeddings.base import Embeddings
+from langchain.schema import BaseRetriever
+from langchain.vectorstores.base import VectorStore
+
+logger = logging.getLogger()
+
+AZURESEARCH_DIMENSIONS = int(
+    os.environ.get("AZURESEARCH_DIMENSIONS", 1536)
+)  # Default to OpenAI's ada-002 embedding model vector size
+
+# Allow overriding field names for Azure Search
+FIELDS_ID = os.environ.get("AZURESEARCH_FIELDS_ID", "id")
+FIELDS_TITLE = os.environ.get("AZURESEARCH_FIELDS_TITLE", "title")
+FIELDS_CONTENT = os.environ.get("AZURESEARCH_FIELDS_CONTENT", "content")
+FIELDS_CONTENT_VECTOR = os.environ.get(
+    "AZURESEARCH_FIELDS_CONTENT_VECTOR", "content_vector"
+)
+FIELDS_TAG = os.environ.get("AZURESEARCH_FIELDS_TAG", "tag")
+FIELDS_METADATA = os.environ.get("AZURESEARCH_FIELDS_TAG", "metadata")
+
+MAX_UPLOAD_BATCH_SIZE = 1000
+
+
+def get_search_client(
+    endpoint: str,
+    key: str,
+    index_name: str,
+    semantic_configuration_name: Optional[str] = None,
+) -> SearchClient:
+    if key is None:
+        credential = DefaultAzureCredential()
+    else:
+        credential = AzureKeyCredential(key)
+    index_client: SearchIndexClient = SearchIndexClient(
+        endpoint=endpoint, credential=credential
+    )
+    try:
+        index_client.get_index(name=index_name)
+    except ResourceNotFoundError:
+        # Fields configuration
+        fields = [
+            SimpleField(
+                name=FIELDS_ID,
+                type=SearchFieldDataType.String,
+                key=True,
+                filterable=True,
+            ),
+            SearchableField(
+                name=FIELDS_TITLE,
+                type=SearchFieldDataType.String,
+                searchable=True,
+                retrievable=True,
+            ),
+            SearchableField(
+                name=FIELDS_CONTENT,
+                type=SearchFieldDataType.String,
+                searchable=True,
+                retrievable=True,
+            ),
+            SearchField(
+                name=FIELDS_CONTENT_VECTOR,
+                type=SearchFieldDataType.Collection(SearchFieldDataType.Single),
+                searchable=True,
+                dimensions=AZURESEARCH_DIMENSIONS,
+                vector_search_configuration="default",
+            ),
+            SearchableField(
+                name=FIELDS_TAG,
+                type=SearchFieldDataType.String,
+                filterable=True,
+                searchable=True,
+                retrievable=True,
+            ),
+            SearchableField(
+                name=FIELDS_METADATA,
+                type=SearchFieldDataType.String,
+                searchable=True,
```

Comment:
A common pattern is to store the filename in the metadata which could be a good candidate for a searchable field. What do you think?

Reply:
Fair enough. It'll be a little noisy since we'd be searching on all text of a json doc, so we might end up matching unexpected stuff. On the other hand, it does help to cases where you're searching for a file name you know about or something like that. Let's start like this, we can always adjust it.